experimental/oauther/oauth2_complete.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package oauther
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+func (o *oAuther) oauth2Complete(w http.ResponseWriter, r *http.Request) {
+	authedUserID := r.Header.Get("Mattermost-User-ID")
+	if authedUserID == "" {
+		o.logger.Debugf("oauth2Complete: reached by non authed user")
+		http.Error(w, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+	code := r.URL.Query().Get("code")
+	if code == "" {
+		o.logger.Debugf("oauth2Complete: reached with no code")
+		http.Error(w, "missing authorization code", http.StatusBadRequest)
+		return
+	}
+	state := r.URL.Query().Get("state")
+
+	storedState, appErr := o.api.KVGet(o.getStateKey(authedUserID))
+	if appErr != nil {
+		o.logger.Warnf("oauth2Complete: cannot get state, err=%s", appErr.Error())
+		http.Error(w, "cannot get stored state", http.StatusInternalServerError)
+		return
+	}
+
+	if string(storedState) != state {
+		o.logger.Debugf("oauth2Complete: state mismatch")
+		http.Error(w, "state does not mach", http.StatusUnauthorized)
+		return
+	}
+
+	userID := strings.Split(state, "_")[1]
+	if userID != authedUserID {
+		o.logger.Debugf("oauth2Complete: authed user mismatch")
+		http.Error(w, "authed user is not the same as state user", http.StatusUnauthorized)
```

Comment:
This is effectively an information leak: we've told the client something they didn't need to know. Can we generalize these error so it's impossible to distinguish the cases (fine to keep logging the particulars).